### PR TITLE
fix(vis-type: scatter): add null check in `useData` hook

### PR DESCRIPTION
**Pull request by oltionchampari**
Closes *list issues numbers here*

### Developer Checklist (Definition of Done)

**Issue**

- [ ] All acceptance criteria from the issue are met
- [ ] Tested in latest Chrome/Firefox

**UI/UX/Vis**

- [ ] Requires UI/UX/Vis review
  - [ ] Reviewer(s) are notified (_tag assignees_)
  - [ ] Review has occurred (_link to notes_)
  - [ ] Feedback is included in this PR
  - [ ] Reviewer(s) approve of concept and design

**Code**

- [ ] Branch is up-to-date with the branch to be merged with, i.e., develop
- [ ] Code is cleaned up and formatted
- [ ] Unit tests are written (frontend/backend if applicable)
- [ ] Integration tests are written (if applicable)

**PR**

- [ ] Descriptive title for this pull request is provided (will be used for release notes later)
- [ ] Reviewer and assignees are defined
- [ ] Add type label (e.g., *bug*, *feature*) to this pull request
- [ ] Add release label (e.g., `release: minor`) to this PR following [semver](https://semver.org/)
- [ ] The PR is connected to the corresponding issue (via `Closes #...`)
- [ ] [Summary of changes](#summary-of-changes) is written

### Summary of changes

- add null checks in useData

### Additional notes for the reviewer(s)

-  
Thanks for creating this pull request 🤗

### CODE REVIEW

File: src/vis/scatter/useData.tsx [hunk 1]
```diff
@@ -195,7 +201,7 @@ export function useData({
       return plots;
     }
 
-    if (splom) {
+    if (splom && value.validColumns[0]) {
```
Comment (thinkh) [inline src/vis/scatter/useData.tsx:204]: @dvmoritzschoefl Is it ok to add the column check here?

File: src/vis/scatter/useData.tsx [hunk 2]
```diff
@@ -93,11 +93,14 @@ export function useData({
...
-            ${value.colorColumn ? `<br />${columnNameWithDescription(value.colorColumn.info)}: ${getLabelOrUnknown(value.colorColumn.resolvedValues[index].val)}` : ''}
-            ${value.shapeColumn && value.shapeColumn.info.id !== value.colorColumn?.info.id ? `<br />${columnNameWithDescription(value.shapeColumn.info)}: ${getLabelOrUnknown(value.shapeColumn.resolvedValues[index].val)}` : ''}`.trim(),
+            ${value.colorColumn ? `<br />${columnNameWithDescription(value.colorColumn.info)}: ${getLabelOrUnknown(value.colorColumn.resolvedValues[index]?.val)}` : ''}
+            ${value.shapeColumn && value.shapeColumn.info.id !== value.colorColumn?.info.id ? `<br />${columnNameWithDescription(value.shapeColumn.info)}: ${getLabelOrUnknown(value.shapeColumn.resolvedValues[index]?.val)}` : ''}`.trim(),
           ),
           marker: {
-            color: value.colorColumn && mappingFunction ? value.colorColumn.resolvedValues.map((v) => mappingFunction(v.val)) : VIS_NEUTRAL_COLOR,
+            color:
+              value.colorColumn && mappingFunction
+                ? value.colorColumn.resolvedValues.map((v) => (v.val ? mappingFunction(v.val) : VIS_NEUTRAL_COLOR))
```
Comment (thinkh) [inline src/vis/scatter/useData.tsx:102]: If `v.val` is unset then use the `VIS_NEUTRAL_COLOR` as fallback.

File: src/vis/scatter/useData.tsx [hunk 3]
```diff
@@ -215,13 +221,16 @@ export function useData({
...
+  ${(value.resolvedLabelColumns ?? []).map((l) => `<br />${columnNameWithDescription(l.info)}: ${getLabelOrUnknown(l.resolvedValues[i]?.val)}`)}
+  ${value.colorColumn ? `<br />${columnNameWithDescription(value.colorColumn.info)}: ${getLabelOrUnknown(value.colorColumn.resolvedValues[i]?.val)}` : ''}
+  ${value.shapeColumn && value.shapeColumn.info.id !== value.colorColumn?.info.id ? `<br />${columnNameWithDescription(value.shapeColumn.info)}: ${getLabelOrUnknown(value.shapeColumn.resolvedValues[i]?.val)}` : ''}`.trim(),
           ),
           ...(isEmpty(selectedList) ? {} : { selectedpoints: selectedList.map((idx) => splom.idToIndex.get(idx)) }),
           marker: {
-            color: value.colorColumn && mappingFunction ? value.colorColumn.resolvedValues.map((v) => mappingFunction(v.val)) : VIS_NEUTRAL_COLOR,
+            color:
+              value.colorColumn && mappingFunction
+                ? value.colorColumn.resolvedValues.map((v) => (v.val ? mappingFunction(v.val) : VIS_NEUTRAL_COLOR))
```
Comment (dvmoritzschoefl) [inline src/vis/scatter/useData.tsx:232]: why is this needed? the mappingFunction should handle undefined/null values (and return the VIS_NEUTRAL_COLOR)
if not we should fix it in ScatterVis (where the mapping function is created)

File: src/vis/scatter/useData.tsx (hunk 1 above)
Comment (dvmoritzschoefl) [reply src/vis/scatter/useData.tsx:204]: it is ok. when splom is valid, validColumns[0] is guaranteed to be there. I added the check for the "strict" mode only since I did not found a better solution to get rid of these errors.

File: src/vis/scatter/useData.tsx [hunk 4]
```diff
@@ -134,15 +137,18 @@ export function useData({
...
+  ${value.shapeColumn && value.shapeColumn.info.id !== value.colorColumn?.info.id ? `<br />${columnNameWithDescription(value.shapeColumn.info)}: ${getLabelOrUnknown(value.shapeColumn.resolvedValues[i]?.val)}` : ''}`.trim(),
           ),
           marker: {
             textfont: {
               color: VIS_NEUTRAL_COLOR,
             },
-            color: value.colorColumn && mappingFunction ? value.colorColumn.resolvedValues.map((v) => mappingFunction(v.val)) : VIS_NEUTRAL_COLOR,
+            color:
+              value.colorColumn && mappingFunction
+                ? value.colorColumn.resolvedValues.map((v) => (v.val ? mappingFunction(v.val) : VIS_NEUTRAL_COLOR))
```
Comment (dvmoritzschoefl) [inline src/vis/scatter/useData.tsx:150]: Same as below. the mapping function should be responsible for returning the vis neutral color

File: src/vis/scatter/useData.tsx (hunk 2 above)
Comment (dvmoritzschoefl) [reply src/vis/scatter/useData.tsx:102]: see same cases below

**Review by dvmoritzschoefl — CHANGES_REQUESTED** [review]

File: src/vis/scatter/useData.tsx (hunk 3 above)
Comment (thinkh) [reply src/vis/scatter/useData.tsx:232]: Yes, good point. Then we must allow null/undefined values in the interface of the props above, right?

```
mappingFunction?: (val: string | number | null | undefined) => string;
```

Comment (dvmoritzschoefl) [reply src/vis/scatter/useData.tsx:232]: yes if its only a type issue this works

File: src/vis/scatter/useData.tsx (hunk 2 above)
Comment (thinkh) [reply src/vis/scatter/useData.tsx]: reverted and added `null | undefined` to function interface.

File: src/vis/scatter/useData.tsx (hunk 3 above)
Comment (thinkh) [reply src/vis/scatter/useData.tsx]: I reverted the changed and adapted the function interface accordingly.

File: src/vis/scatter/useData.tsx (hunk 4 above)
Comment (thinkh) [reply src/vis/scatter/useData.tsx]: Reverted.

**Review by dvmoritzschoefl — APPROVED** [review]
